gobblin-core-base/src/main/java/gobblin/writer/SequentialAndTTLBasedBatchAccumulator.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package gobblin.writer;
+
+import java.util.LinkedList;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.Future;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.util.concurrent.Futures;
+import com.typesafe.config.Config;
+
+import gobblin.util.ConfigUtils;
+
+
+/**
+ * Sequential and TTL based accumulator
+ * A producer can add a record to this accumulator. It generates a batch on the first record arrival. All subsequent records
+ * are added to the same batch until a batch size limit is reached. {@link BufferedAsyncDataWriter} keeps iterating available
+ * batches from this accumulator, all completed batches (full sized) will be popped out one by one but an incomplete batch
+ * keeps in the deque until a TTL is expired.
+ */
+
+public abstract class SequentialAndTTLBasedBatchAccumulator<D> extends BatchAccumulator<D> {
```

Comment:
`SequentialAndTTLBasedBatchAccumulator` is too verbose. `SequentialBatchAccumulator` is good enough. Usually, asynchronization comes together with timeout.